### PR TITLE
Subscription Change Notifications

### PR DESCRIPTION
**Pull request by LinsonI**

### CODE REVIEW

File: include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h [hunk 1]
```diff
@@ -133,53 +134,109 @@ class USubscriptionClientDb : public UListener {
...
+         */
+        UCode registerForNotifications(const UUri &topicUri,
+                                       const notifyFunc &func ) {
+            std::string topic;
+            if (!topicUri.SerializeToString(&topic)) {
+                return UCode::INTERNAL;
+            }
 
-        //    notifyMap_[serUri] = func;
+            notifyMap_[topic] = func;
```
Comment (MelamudMichael) [inline include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h:155]: check if already exists, return an error if yes

File: include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h [hunk 2]
```diff
@@ -133,53 +134,109 @@ class USubscriptionClientDb : public UListener {
...
+         * @param topicUri - Topic to be registered.
+         * @param func - Callback function to notify.
+         *
+         * @return Return OK if success.
+         */
+        UCode registerForNotifications(const UUri &topicUri,
+                                       const notifyFunc &func ) {
+            std::string topic;
+            if (!topicUri.SerializeToString(&topic)) {
+                return UCode::INTERNAL;
```
Comment (MelamudMichael) [inline include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h:152]: print error

**Comment by linsonittoop** [comment]
Fixed review comments

File: src/zenohUTransport.cpp [hunk 3]
```diff
@@ -175,7 +176,7 @@ UStatus ZenohUTransport::send(const UUri &uri,
         status.set_code(UCode::UNAVAILABLE);
         return status;
     }
-    
+
     /* determine according to the URI is the send is and RPC response or a regular publish */
     if (false == isRPCMethod(uri.resource())) {
         // if (UCode::OK != getPublisherStatus(uri)) {
```
Comment (MelamudMichael) [inline src/zenohUTransport.cpp:182]: why is this commented out?

File: include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h [hunk 4]
```diff
@@ -95,34 +70,31 @@ class USubscriptionClientDb : public UListener {
...
+            return subStatusMap_[topic];
         }
 
         UCode getPublisherStatus(const UUri &uri) {
 
             // auto u = Temp::buildTopic(uri);
-            
+
             // std::string serUri;
             // if (!u.SerializeToString(&serUri)) {
```
Comment (MelamudMichael) [inline include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h:99]: why this is commented out?

**Review by MelamudMichael — CHANGES_REQUESTED** [review]

File: include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h [hunk 5]
```diff
@@ -133,61 +105,193 @@ class USubscriptionClientDb : public UListener {
...
+         * @return Return UStatus.
+         */
+        UStatus onReceive(const UUri &uri,
+                          const UPayload &payload,
+                          const UAttributes &attributes) const override {
+            UStatus status;
+            status.set_code(UCode::OK);
+            std::string str = uprotocol::uri::LongUriSerializer::serialize(uri);
+            std::string subUri = uprotocol::uri::LongUriSerializer::serialize(uSubUpdatesUri);
+            if (str == subUri) {
```
Comment (MelamudMichael) [inline include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h:283]: As you registered the OnRecive for the following URI 
("/core.usubscription/3/subscriptions#Update");

You will always receive message on this specific URI , so this check is not needed.
Instead you need to check the payload , as it is the one that actually contains the URI that you are getting the Update for

File: include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h [hunk 6]
```diff
@@ -133,61 +105,193 @@ class USubscriptionClientDb : public UListener {
...
+                return UCode::INTERNAL;
+            }
+
+            subStatusMap_[subscribedTopic] = update.status().state();
+
+            auto it = notifyMap_.find(subscribedTopic);
+            if (notifyMap_.end() != it) {
+                notifyFunc notify = it->second;
+                if (nullptr != notify) {
+                    notify(update.status());
```
Comment (MelamudMichael) [inline include/uprotocol-cpp-ulink-zenoh/usubscription/internal/uSubscriptionClientDb.h:191]: Please check the return value

File: examples/subscription/src/main_subscription.cpp [hunk 7]
```diff
@@ -35,164 +35,263 @@ using namespace uprotocol::uuid;
...
+    virtual void test() = 0;
 
-    request.set_allocated_topic(&uri);
-    
-    return request;
-}
+protected:
 
-UnsubscribeRequest buildUnsubscribeRequest(UUri uri) {
+    int init() {
```
Comment (MelamudMichael) [inline examples/subscription/src/main_subscription.cpp:82]: the convention is to use UCode

File: examples/subscription/src/main_subscription.cpp [hunk 8]
```diff
@@ -35,164 +35,263 @@ using namespace uprotocol::uuid;
 using namespace uprotocol::v1;
```
Comment (MelamudMichael) [inline examples/subscription/src/main_subscription.cpp:35]: You are missing the test that will check the actual notification logic.
For example - 
1. send subscription request for a remote topic
2. server returns SUBSCRIBE_PENDING
3. As we dont have remote subscriptions support yet, the server should wait for one second, and send notification SUBSCRIBED (this part should be simulated in the server implementation)

**Review by MelamudMichael — CHANGES_REQUESTED** [review]

**Pull request closed without merging**